### DISCUSSION TITLE
Get class and method info from the cursor position/selection

### DISCUSSION THREAD

**Discussion by tonka3000** [microsoft/vscode-discussions#637] [Extension Development QnA] [answered]
Hi,

is it possible to get class/method information from the current cursor position or selection which is provided by another extension via the VSCode API?

Right now I would only be able when I write a small parser around the cursor position/selection.

## Example

Having a C++ code, where the class and method are already highlighted. I would need the names and the type to be able to generate a URL for a documentation system

**Comment by r3m0t** [comment]
Hi, you can use `vscode.executeDocumentSymbolProvider` to get the same information the `Outline` panel would use. That's the most generic method. It will recalculate it so use caching and consider requesting the information before you need it.

https://code.visualstudio.com/api/references/commands

If it's specific to one language you can look at the extension for the language and whether it exposes an API for other extensions, or use something like tree-sitter.

**Comment by alefragnani** [comment, marked as answer]
Hi,

There is no such API in VS Code. And, as commented by @r3m0t , your best bet for a _generic_ approach would be to use `vscode.executeDocumentSymbolProvider`, to get the symbols within the current file.

But even using `vscode.executeDocumentSymbolProvider`, you still  have to _interpret_ the results, depending on the language opened in the editor. This is necessary because each language has its own _server_ and depends on each server to provide the values used in the outline.

For instance, TypeScript ang Go returns just the function name in the `Symbol.name` attribute.  C# and Java, on the other hand, returns the function signature (function name and its params). You must parse the text to correctly detect the function name. To better understand the differences, take a look at the Outline view for different languages

Hope this helps

**Reply by StErMi** [reply]
I have a stupid question, who's in charge of the "server" you are talking about? If `executeDocumentSymbolProvider` doesn't return the whole signature and the needed information, who should update it?

**Reply by alefragnani** [reply]
Depends on the language. It may be built in on VS Code, for languages like JS, TS and even Markdown, or via some extension that you must have installed, like C#, Go, Java and Python. If you don’t see info in the Outline view, you don’t have a language extension that supports it. If you see “incomplete” information, it means the language extension you have installed doesn’t provide the whole information. In this case, you must contact the extension author.

BTW, simple _syntax highlighting_ extension isn’t enough. The extension must provide the `Go to Symbol` feature, as it is where the `executeDocumentSymbolProvider` comes from. 

Ps. this is not, in any way, a stupid question. It is in fact, a really important one, specially for extension developers that needs any “language parser” related information.

Hope this helps

**Comment by tonka3000** [comment]
Thanks for your answers @r3m0t @alefragnani

I will checkout `vscode.executeDocumentSymbolProvider`. Luckly I only have to support C++ for now, but @alefragnani hint about different return values of the different symbol providers is quite helpful to know.

I will keep you posted here when I've tried it.

**Reply by tonka3000** [reply]
I tried `vscode.executeDocumentSymbolProvider`, but C++ extension only give the outer function back (also in the outline).
I check the C++ extension for an API, but sadly they don't have the info for now. I will raise an issue in the repo.

**Reply by alefragnani** [reply]
If the C++ extension that you use doesn’t provide all functions in the Outline, then `vscode.executeDocumentSymbolProvider` won’t work for you, unfortunately. 

You are right asking the extension authors to improve the extension. Maybe they could simply improve the Outline view, but exposing an API is an alternative. Or, you could look for other extensions that provides the information you need (of course, if both extensions doesn’t clash with each other).

Good luck.
